File: problem_12/problem_12.py
import math
from functools import reduce


def triangle_number(before: int, n: int) -> int:
    return before + n


def factors(n: int) -> int:
    step = 2 if n % 2 else 1
    return len(
        set(
            reduce(
                list.__add__,
                (
                    [i, n // i]
                    for i in range(1, int(math.sqrt(n)) + 1, step)
                    if n % i == 0
                ),
            )
        )
    )


n = 1
before = 0
while True:
    tn = triangle_number(before, n)
    print(tn)
    if factors(tn) > 500:
        print("Result:", tn)
        break
    before = tn
    n += 1

File: problem_12/problem_12.go
package main

import (
	"fmt"
	"math"
)

func triangle_numbers(before, n int) int {
	return before + n
}

func factors(n int) int {
	step := 2
	if n%2 == 0 {
		step = 1
	}
	m := map[int]bool{}
	for i := 1; i <= int(math.Sqrt(float64(n))); i += step {
		if n%i == 0 {
			m[i] = true
			m[n/i] = true
		}
	}
	return len(m)
}

func main() {
	var n, before int = 1, 0
	for {
		tn := triangle_numbers(before, n)
		fmt.Println(tn)
		if factors(tn) > 500 {
			fmt.Println("Result:", tn)
			break
		}
		before = tn
		n += 1
	}
}
